workflow/train_test_split.py

```diff
@@ -0,0 +1,165 @@
+from collections.abc import Iterable
+import numpy as np
+import pandas as pd
+import pytest
+
+
+def train_test_split(df, key, test_size=0.2, random_state=None, stratify=None):
+
+    original_random_state = np.random.get_state()
+    np.random.seed(random_state)
+
+    if stratify is None:
+        test_df = get_test_split(df, key, test_size)
+    else:
+        test_df = multilabel_stratified_split(df, key, stratify, test_size)
+
+    train_df = df[~df[key].isin(test_df[key])]
+    np.random.set_state(original_random_state)
+
+    return train_df, test_df
+
+
+def get_test_split(df, key, test_size, resolution=100000):
+    salt = np.random.randint(0, resolution)
+    split_key = df[key].apply(lambda x: ((hash(hash(x) + salt) % resolution)))
+    return df[split_key < resolution * test_size]
+
+
+def multilabel_stratified_split(df, key, stratify, test_size):
+
+    test_dfs = []
+    for label in get_labels(df, stratify):
+        if type(stratify) is str:
+            df_part = df[df[stratify] == label]
+        else:
+            df_part = df[df[label].notnull()]
+        test_dfs.append(get_test_split(df_part, key, test_size))
+        df = df[~df[key].isin(df_part[key])]
+    assert len(df) == 0, f'Unable to stratify on {stratify}'
+
+    return pd.concat(test_dfs)
+
+
+def get_labels(df, stratify):
+
+    if type(stratify) is str:
+        return sorted(df[stratify].unique())
+    else:
+        if not isinstance(stratify, Iterable):
+            raise ValueError(
+                f'Argument stratify must be str or iterable, was {stratify}'
+            )
+        return sorted(stratify)
```

Comment:
What is this other type of stratify you are supporting? When would you use it?

Reply:
I replied above but can clarify. If stratify is type str it will do a normal stratify, if it is an iterable it will do a multilabel stratification. Since the algorithm works the same way for both cases I let it sit in the same function, but maybe it should be split up?

Reply:
I think its clear from the code :) perhaps  something that could be made better later, but its most definitely good enough to pull into master

Reply:
Why would you want to input a list as stratify?

Reply:
Ah it's when you have one column per label in the dataframe